mgr/dashboard: add restful api for creating crush rule with type of 'erasure'

Serrini opened the pull request:
Fixes: https://tracker.ceph.com/issues/66490
Signed-off-by: sunlan <sunlan@asiainfo.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nizamial09 commented on src/pybind/mgr/dashboard/controllers/crush_rule.py:
```diff
@@ -47,6 +47,16 @@ def create(self, name, root, failure_domain, device_class=None):
         }
         CephService.send_command('mon', 'osd crush rule create-replicated', **rule)
 
+    @RESTController.Resource('POST')
+    def create_erasure(self, name, profile, failure_domain, device_class):
+        rule = {
+            'name': name,
+            'profile': profile,
+            'type': failure_domain,
+            'class': device_class
+        }
+        CephService.send_command('mon', 'osd crush rule create-erasure', **rule)
+
```
Instead of creating a separate API for erasure, can we combine it along with the existing one. It looks like both of them are almost similar with one exception. So we can toggle them based on `replication` or `erasure` type. That's the standard we have across the dashboard and can keep the APIs minimized.

nizamial09 reviewed (commented):
Thank you for the contribution @Serrini , this looks good to me but I added a suggestion there. Also can you add[ an api test ](https://github.com/ceph/ceph/blob/main/qa/tasks/mgr/dashboard/test_crush_rule.py)here to validate the change?

Serrini commented:
> Thank you for the contribution @Serrini , this looks good to me but I added a suggestion there. Also can you add[ an api test ](https://github.com/ceph/ceph/blob/main/qa/tasks/mgr/dashboard/test_crush_rule.py)here to validate the change?

@nizamial09  Thank you for your suggestion. I have merged these two interfaces and added testing code in the 'test_crush_rule.py'. These changes will not affect the original 'replication' interface.

nizamial09 reviewed (approved):
thank you @Serrini lgtm! let's wait for the tests to pass and then i'll merge it.

nizamial09 commented on src/pybind/mgr/dashboard/controllers/crush_rule.py:
```diff
@@ -37,16 +37,23 @@ def get(self, name):
-        }
-        CephService.send_command('mon', 'osd crush rule create-replicated', **rule)
-
+    def create(self, name, failure_domain, device_class, root=None, profile=None, pool_type=None):
```
maybe we can default it to `replication` by default?
```suggestion
    def create(self, name, failure_domain, device_class, root=None, profile=None, pool_type='replication'):
```
```suggestion
    def create(self, name, failure_domain, device_class, root=None, profile=None, pool_type=None):
```

nizamial09 commented:
There are 2 [test failures](https://jenkins.ceph.com/job/ceph-pull-requests/137648/)

[Test Result](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/) (2 failures / +2)
[projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_lint](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_lint/)
[projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_openapi](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_openapi/)

for fixing the openapi failure, you can run `tox -e openapi-fix` from the `mgr/dashboard` folder.
https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#rest-api-documentation

nizamial09 commented:
and the[ api test](https://jenkins.ceph.com/job/ceph-api/76605/consoleFull#1303174777c212b007-e891-4176-9ee7-2f60eca393b7) is failing on
````
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
    return handler(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/cherrypy/_cpdispatch.py", line 54, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/_base_controller.py", line 263, in inner
    ret = func(*args, **kwargs)
  File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/_rest_controller.py", line 193, in wrapper
    return func(*vpath, **params)
TypeError: CrushRule.create() missing 1 required positional argument: 'device_class'
2024-06-26T11:15:10.562+0000 7fe41642b640  0 [dashboard ERROR request] [::ffff:172.21.15.15:59558] [POST] [500] [0.006s] [admin] [513.0B] /api/crush_rule
2024-06-26T11:15:10.562+0000 7fe41642b640  0 [dashboard ERROR request] [b'{"status": "500 Internal Server Error", "detail": "The server encountered an unexpected condition which prevented it from fulfilling the request.", "request_id": "77f80ca7-b03a-4981-bda1-528b56bf4957"}                                                                                                                                                                                                                                                                                                                        ']
```

Serrini commented:
jenkins test API

Serrini commented:
jenkins test make check

Serrini commented:
> 有 2 次[make check 失败](https://jenkins.ceph.com/job/ceph-pull-requests/137648/)
> 
> [测试结果（ ](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/)[2](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_openapi/)次失败/+2）projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_lint [projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_openapi](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_lint/) [](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_openapi/)
> 
> 为了修复 openapi 故障，您可以`tox -e openapi-fix`从`mgr/dashboard`文件夹运行。https [://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#rest-api-documentation](https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#rest-api-documentation)

Serrini commented:
> 并且[API 测试](https://jenkins.ceph.com/job/ceph-api/76605/consoleFull#1303174777c212b007-e891-4176-9ee7-2f60eca393b7)失败
> 
> ```
>   File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
>     return handler(*args, **kwargs)
>   File "/usr/lib/python3/dist-packages/cherrypy/_cpdispatch.py", line 54, in __call__
>     return self.callable(*self.args, **self.kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/_base_controller.py", line 263, in inner
>     ret = func(*args, **kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/_rest_controller.py", line 193, in wrapper
>     return func(*vpath, **params)
> TypeError: CrushRule.create() missing 1 required positional argument: 'device_class'
> 2024-06-26T11:15:10.562+0000 7fe41642b640  0 [dashboard ERROR request] [::ffff:172.21.15.15:59558] [POST] [500] [0.006s] [admin] [513.0B] /api/crush_rule
> 2024-06-26T11:15:10.562+0000 7fe41642b640  0 [dashboard ERROR request] [b'{"status": "500 Internal Server Error", "detail": "The server encountered an unexpected condition which prevented it from fulfilling the request.", "request_id": "77f80ca7-b03a-4981-bda1-528b56bf4957"}                                                                                                                                                                                                                                                                                                                        ']
> ```
> ```

hello. I've fixed this issue and

Serrini commented:
> There are 2 [make check failures](https://jenkins.ceph.com/job/ceph-pull-requests/137648/)
> 
> [Test Result](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/) (2 failures / +2) [projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_lint](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_lint/) [projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_openapi](https://jenkins.ceph.com/job/ceph-pull-requests/137648/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_openapi/)
> 
> for fixing the openapi failure, you can run `tox -e openapi-fix` from the `mgr/dashboard` folder. https://docs.ceph.com/en/latest/dev/developer_guide/dash-devel/#rest-api-documentation

hello. I've fixed it.

Serrini commented:
jenkins test make check

Serrini commented:
> and the[ api test](https://jenkins.ceph.com/job/ceph-api/76605/consoleFull#1303174777c212b007-e891-4176-9ee7-2f60eca393b7) is failing on
> 
> ```
>   File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
>     return handler(*args, **kwargs)
>   File "/usr/lib/python3/dist-packages/cherrypy/_cpdispatch.py", line 54, in __call__
>     return self.callable(*self.args, **self.kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/_base_controller.py", line 263, in inner
>     ret = func(*args, **kwargs)
>   File "/home/jenkins-build/build/workspace/ceph-api/src/pybind/mgr/dashboard/controllers/_rest_controller.py", line 193, in wrapper
>     return func(*vpath, **params)
> TypeError: CrushRule.create() missing 1 required positional argument: 'device_class'
> 2024-06-26T11:15:10.562+0000 7fe41642b640  0 [dashboard ERROR request] [::ffff:172.21.15.15:59558] [POST] [500] [0.006s] [admin] [513.0B] /api/crush_rule
> 2024-06-26T11:15:10.562+0000 7fe41642b640  0 [dashboard ERROR request] [b'{"status": "500 Internal Server Error", "detail": "The server encountered an unexpected condition which prevented it from fulfilling the request.", "request_id": "77f80ca7-b03a-4981-bda1-528b56bf4957"}                                                                                                                                                                                                                                                                                                                        ']
> ```
> ```

@nizamial09 hello. The API test passed.

Serrini commented:
jenkins dashboard tests

Serrini commented:
jenkins ceph dashboard tests

Serrini commented:
@nizamial09 @afreen23 @aaSharma14  Hello, here's a workflow waiting for approval and what can I do for it.

nizamial09 commented:
@Serrini sorry for missing out on this. Everything looks good now, thanks. Just one last thing, can you squash all the commits into one? And once that is done, I'll merge the PR.

Serrini commented:
jenkins test make check

Serrini commented:
jenkins test make check

nizamial09 commented:
jenkins test make check

Serrini commented:
@nizamial09 Hello, thank you for your help. The process seems to be working fine now

nizamial09 commented:
thank you again for the contribution. @Serrini. The PR is merged. I have updated the tracker and marked it for squid backport. Let me know if it needs to be in any other branch than squid.
https://tracker.ceph.com/issues/66490
